pkg/kubelet/heartbeat.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/client-go/util/clock"
+)
+
+const (
+	// These constants should be reviewed and updated very carefully.
+	// nodeStatusUpdateRetry the normal default for how many times kubelet retries when posting node status failed.
+	// this constant is a convention that is expected to be depended on externaly by other components.
+	NodeStatusUpdateRetry      = 5
+	nodeRegistrationMaxBackoff = 7 * time.Second
+
+	// Less critical constants subject to change.
+	nodeStatusBurstRetry = 5 * time.Millisecond
+)
+
+// This module externalizes logic for heartbeating and retrying of communication with the API server.
+// It is intended to externalize functionality that other components (like the controller) may depend on
+
+// registerInitiallyWithInfiniteRetry continually runs a function until it succeeds, backing off
+// using specific timings we set for initial registration flow.  Specifically designed for initial registration,
+//  externalized to make timing policy easily maintainable.
+func registerInitiallyWithInfiniteRetry(theClock clock.Clock, register func() bool, message string) error {
```

Comment:
Eventually, yes.  

But as of now, 

- this logic needs to be preserved for this PR, and once we decide how to evolve the backoffs, we likely can use more generic functionality.  
- And even still, the heartbeat module IMO lives on and wraps those utilities so that the core functionality of node communication is insulated and easy to reason about.